File: nodejs/Permutation.js
"use strict";

/**
 * Generator function for Heap's algorithm
 * new Permutation(n).next().value returns next permutation of n elements
 * @param {Number} n Permutation elements
 */
function* Permutation(n){
    const permutation = new Array(n);
    const idx = new Array(n);
    let _ = null;

    for(let i=0;i<n;++i) {
        permutation[i] = i;
        idx[i]=0;
    }

    yield permutation;

    for(let i=0;i<n;) {
        if (idx[i] < i) {
            const swap = i % 2 * idx[i];
            _ = permutation[swap];
            permutation[swap] = permutation[i];
            permutation[i] = _;
            yield permutation;

            idx[i]++;
            i = 1;
        }
        else
            idx[i++] = 0;
    }
}

module.exports=Permutation;

File: c/permutation.cc
#include "permutation.h"
#include <iostream>
#include <time.h>
#include <cstdlib>

using namespace std;

Permutation::Permutation(int n_){
    i=1;
    n=n_;
    idx=(int*)malloc(sizeof idx * n);
    permutation=(_INT*)malloc(sizeof permutation * n);
    for(int i=0;i<n;i++){
        permutation[i]=i;
        idx[i]=0;
    }
}

_INT* Permutation::next(){
    for (;i<n;){
        if(idx[i]<i){
            int swap = i%2*idx[i];
            int tmp = permutation[swap];
            permutation[swap] = permutation[i];
            permutation[i] = tmp;
            idx[i]++;
            i = 1;
            return permutation;
        }
        else
            idx[i++] = 0;
    }
    return NULL;
}
